lib/protocol/protocol.go

```diff
@@ -867,26 +870,29 @@ func (c *rawConnection) shouldCompressMessage(msg message) bool {
 // BEP message is sent before terminating the actual connection. The error
 // argument specifies the reason for closing the connection.
 func (c *rawConnection) Close(err error) {
-	c.sendCloseOnce.Do(func() {
-		done := make(chan struct{})
-		timeout := time.NewTimer(CloseTimeout)
-		select {
-		case c.closeBox <- asyncMessage{&Close{err.Error()}, done}:
+	// Close might be called from a method that is called from within
+	// dispatcherLoop, resulting in a deadlock.
+	// The sending above must happen before spawning the routine, to prevent
+	// the underlying connection from terminating before sending the close msg.
+	timeout := time.NewTimer(CloseTimeout) // To prevent a testing race
+	go func() {
+		c.sendCloseOnce.Do(func() {
+			close(c.sendStop)
+			done := make(chan struct{})
 			select {
-			case <-done:
+			case c.closeBox <- asyncMessage{&Close{err.Error()}, done}:
+				select {
+				case <-done:
+				case <-timeout.C:
+				case <-c.closed:
```

Comment:
This is confusing. We await for something but there are no calls after this, why do we care on waiting for this subselect?

Reply:
We are waiting for `done`/the send to go through (or timeout or connection being closed under our feat) before calling `c.internalClose` further down. While it doesn't make a difference we could also return on the `c.closed` cases (`c.internalClose` is a noop if `c.closed` is closed).